lib/plugins/threaded-signing/plugin.c

```diff
@@ -154,19 +153,22 @@ signing_worker_thread(void *user_data)
   while (self->is_running) {
     if (self->input_buffer_idx > 0) {
       // Get the oldest hash from the input buffer
-      // Copy the hash to |signature_info| and start signing. In principle, it is now possible to
-      // prepare for a new hash.
+      // Copy the hash to |signature_info| and start signing.
       assert(self->hash_size == self->signature_info->hash_size);
       assert(self->signature_info->hash);
       memcpy(self->signature_info->hash, self->input_buffer[0], self->hash_size);
 
       uint8_t *tmp = self->input_buffer[0];
       int j = 0;
-      while (self->input_buffer[j] != NULL && j < MAX_BUFFER_LENGTH) {
+      while (self->input_buffer[j] != NULL && j < MAX_BUFFER_LENGTH - 1) {
         self->input_buffer[j] = self->input_buffer[j + 1];
         j++;
       }
-      self->input_buffer[j - 1] = tmp;
+      if (j < MAX_BUFFER_LENGTH - 1) {
```

Comment:
You can get rid of this if-else and replace it with ```self->input_buffer[j] = tmp;``` if you in the wile loop use ```self->input_buffer[j+1] != NULL```